Various improvements to K frontend

dwightguth opened the pull request:
At a high level:

* refactor the parser generator so that modules other than ID can have separate syntax when parsing programs (needed for upcoming substitution PR)
* fixes a bug where macros were not expanded if they appear in the initial configuration
* adds support for `#Or` in the lhs of rules which is interpreted as an or-pattern a la OCAML
* adds new operators :=K and :/=K whose mathematical meaning is \in(...) and \not(\in(...))) and which is compiled efficiently by the frontend down to pattern matching on a fresh function symbol.
* fixes a bug in #266 which caused the incorrect guice jar to be pulled in as a dependency of some of our other guice dependencies.

Commit:
generic -PROGRAM-PARSING not just for ID

Commit:
expand macros in llvm backend

Commit:
expand macros in haskell backend

Commit:
support or patterns in K frontend

Commit:
fix associativity of #Or and #And

Commit:
add tests for lambda

Commit:
document new behavior of ==K and =/=K

Commit:
allow unbound vars on rhs of ==K and =/=K

Commit:
resolve ==K and =/=K to lambdas

Commit:
change syntax of pattern matching operator

Commit:
fix bug in classpath

traiansf commented on k-distribution/include/builtin/kast.k:
```diff
@@ -73,18 +73,16 @@ module KSEQ-SYMBOLIC
+  syntax KBott ::= "#True" [klabel(#True), symbol]
+  syntax KBott ::= "#False" [klabel(#False), symbol]
+  syntax KBott ::= "#Not" "(" K ")" [klabel(#Not), symbol, mlOp, poly(0, 1)]
+                 | "#Ceil" "(" K ")" [klabel(#Ceil), symbol, mlOp, poly(0, 1)]
```
Does this mean that I'll be able to use (all) the ML connectives when writing K definitions?

traiansf reviewed (approved):
Looks good to me.

ehildenb commented:
I think we need to have a discussion about the surface syntax of ML that's exposed to users. @yzhang90 actually needs some other constructs for his bounded model checking, which probably we need an `#Always` and an `#Eventually`. But I would prefer if surface K exposed prettier versions of the operators, eg. `_/\_` for ML AND, and `_\/_` for ML OR, or `[]_` for Always (so that we're consistent with CTL/LTL).

traiansf commented:
Always and eventually are not even present in KORE :-)
But I fully agree with you;  I think that claims in specs, for example, should be written using the appropriate modal operators instead of the plain rewrite.

ehildenb reviewed (approved):
Should be OK, since as we discussed this is about just lifting ML operators to surface syntax, we can add versions of the operators which do some clever/userful things down the road.

Commit:
update syntax of ml primitives in K

radumereuta commented on kernel/src/main/java/org/kframework/compile/checks/ComputeUnboundVariables.java:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2017-2019 K Team. All Rights Reserved.
+// Copyright (c 2017-2019 K Team. All Rights Reserved.
```
Is that right?

Commit:
fix copyright

Commit:
fix indent

Commit:
fix bug with formatter and polymorphic associative klabels

Commit:
Merge branch 'master' into fixes

Commit:
update llvm submodule

Commit:
or patterns in ocaml backend

Commit:
add tests for or patterns

Commit:
add tests for or patterns

Commit:
fix syntax again

Commit:
use only minimal productions when creating ambiguities during unparsing

This solves an OutOfMemoryError when unparsing large associative lists
of polymorphic productions, which were exploding in the size of the tree
during unparsing. We should never need to choose something that is not
minimal with respect to the overload relation because a minimal
production can always go anywhere a production would go that is greater
than it.

Commit:
check well sortedness when constructing tree nodes

Commit:
fix bug where extension module was not used to unparse collection items

Commit:
update test outputs

Commit:
update to k exercises

dwightguth commented:
This should hopefully finally be passing tests. I ended up having to take a dive into unparsing in order to get the tutorial working again with my changes, so @ehildenb and @traiansf should probably re-review.

Commit:
fix bug with empty amb node

Commit:
Merge branch 'master' into fixes
